tests/Microsoft.Bot.Builder.TestBot/DialogBotBase.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Dialogs;
+
+namespace Microsoft.Bot.Builder.TestBot
+{
+    /// <summary>
+    /// This base class can be used for testing on a single node using something like MemoryStore
+    /// because it uses a lock to make sure activities that arrive at the same time get correctly
+    /// handled by the same Dialog.
+    /// Note the lock here is very wide - all instances regardless of conversation - so this is only
+    /// useful for testing your dialog on a single box.
+    /// </summary>
+    public class DialogBotBase : IBot
+    {
+        private SemaphoreSlim _semaphore;
+        private IBotStoreManager _storeManager;
+        private string _rootDialogName;
+
+        public DialogBotBase(TestBotAccessors accessors, string rootDialogName)
+        {
+            // create the DialogSet from accessor
+            Dialogs = new DialogSet(accessors.ConversationDialogState);
+
+            // a semaphore to serialize access to the bot state
+            _semaphore = accessors.SemaphoreSlim;
+
+            // the state to attempt to save at the end of every turn
+            _storeManager = accessors.StoreManager;
+
+            // the name of the root dialog this bot runs
+            _rootDialogName = rootDialogName;
+        }
+
+        protected DialogSet Dialogs { get; private set; }
+
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken)
+        {
+            // We only want to pump one activity at a time through the state.
+            // Note the state is shared across all instances of this IBot class so we
+            // create the semaphore globally with the accessors.
+            try
+            {
+                // Assuming we are using BotState and MemoryStore we need some locking.
+                await _semaphore.WaitAsync();
```

Comment:
Pass in the cancellationToken. 

https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.waitasync?view=netframework-4.7.2#System_Threading_SemaphoreSlim_WaitAsync_System_Threading_CancellationToken_